### PR TITLE
fix(dvm2.0): M-01 - Inaccurate function results

### PR DESCRIPTION
**Pull request by md0x**
OZ identified the following issue:

```
M-01 Inaccurate function results

The following functions return information about the state of the VotingV2 contract:
getPendingRequests : returns an array of pending requests to be voted on in the
current round.
currentActiveRequests : returns a boolean to indicate whether there are any active
requests.
_getPriceOrError : returns a boolean to indicate success or failure to retrieve a price
for a price request, a price (if applicable), and a string error (if applicable).
getNumberOfPriceRequests : returns the number of pending price requests and the
number of resolved price requests.
However, these functions can return results that are inconsistent with the true state of the
VotingV2 contract. The functions getPendingRequests , currentActiveRequests ,
and _getPriceOrError use the _getRequestStatus function to determine whether a
request is still active or not. _getRequestStatus may determine that a price request has
been resolved, even if it has not yet been processed via the
_processResolvablePriceRequests function in the contract. Additionally, the
_getRequestStatus function does not take into consideration whether pending price
requests will be deleted. The getNumberOfPriceRequests function simply returns the
length of the pendingPriceRequestsIds and resolvedPriceRequestIds arrays,
which can be outdated if _processResolvablePriceRequests has not yet been called for
the current round.
```

The solution in this PR is to update the [_getRequestStatus()](https://github.com/UMAprotocol/protocol/blob/md0x/dvm-audit-m01/packages/core/contracts/data-verification-mechanism/implementation/VotingV2.sol#L1041) to consider the case where the price request should be deleted because of too many rolls. This update fixes all the issues allowing us to remove the `...PostUpdate()` functions except for [getNumberOfPriceRequestsPostUpdate()](https://github.com/UMAprotocol/protocol/blob/0a6d6bb34935bd6ec0f435bc622aefbfb22afd9b/packages/core/contracts/data-verification-mechanism/implementation/VotingV2.sol#L610) for which we have kept both versions of it for convenience: `getNumberOfPriceRequestsPostUpdate` and `getNumberOfPriceRequests`. We have documented the latter properly.

### CODE REVIEW

File: packages/core/contracts/data-verification-mechanism/implementation/VotingV2.sol [hunk 1]
```diff
@@ -520,22 +520,22 @@ contract VotingV2 is Staker, OracleInterface, OracleAncillaryInterface, OracleGo
      * @notice Gets the requests that are being voted on this round.
      * @return pendingRequests array containing identifiers of type PendingRequestAncillaryAugmented.
      */
-    function getPendingRequests() public override returns (PendingRequestAncillaryAugmented[] memory) {
+    function getPendingRequests() public view override returns (PendingRequestAncillaryAugmented[] memory) {
```
Comment (md0x) [inline packages/core/contracts/data-verification-mechanism/implementation/VotingV2.sol:523]: This missing view is reported in `N-02 Functions can be marked as view`

File: packages/core/contracts/data-verification-mechanism/implementation/VotingV2.sol [hunk 2]
```diff
@@ -1065,13 +1061,22 @@ contract VotingV2 is Staker, OracleInterface, OracleAncillaryInterface, OracleGo
         view
         returns (RequestStatus)
     {
-        if (priceRequest.lastVotingRound == 0) return RequestStatus.NotRequested;
-        else if (priceRequest.lastVotingRound < currentRoundId) {
+        if (priceRequest.lastVotingRound == 0) {
+            return RequestStatus.NotRequested;
+        }
```
Comment (chrismaree) [inline packages/core/contracts/data-verification-mechanism/implementation/VotingV2.sol:1066]: small formatting nits that some of. these can be done on one line.

File: packages/core/contracts/data-verification-mechanism/implementation/VotingV2.sol [hunk 3]
```diff
@@ -1116,4 +1121,20 @@ contract VotingV2 is Staker, OracleInterface, OracleAncillaryInterface, OracleGo
...
+    // Returns the actual roll count of a request. This is the roll count plus the number of rounds that have passed
+    // since the last voting round.
+    function _getActualRollCount(PriceRequest storage priceRequest, uint32 currentRoundId)
+        private
+        view
+        returns (uint32)
+    {
+        if (currentRoundId <= priceRequest.lastVotingRound) return priceRequest.rollCount;
+        return priceRequest.rollCount + currentRoundId - priceRequest.lastVotingRound;
+    }
```
Comment (chrismaree) [inline packages/core/contracts/data-verification-mechanism/implementation/VotingV2.sol:1139]: there are no gas implications from pulling these out to private/internal functions, right?

Comment (md0x) [reply packages/core/contracts/data-verification-mechanism/implementation/VotingV2.sol:1139]: TL;DR: Overall, the new gas consumptions seem to have slightly lower average gas consumption than the old ones, but the maximum gas consumption is either the same or slightly higher in some cases.

| Function | min  | max | avg |
|--|--|--|--|
| NEW : processResolvablePriceRequests | 36221 |10033176 | 398914|
| OLD: processResolvablePriceRequests | 36221 |10000676 | 408182|
| NEW: requestPrice | 60640 |176259 | 160237|
| OLD: requestPrice | 60662 |176281 | 160922|
| NEW: updateTrackers | 55614 |850173 | 127541|
| OLD: updateTrackers | 55636 |850195 | 127484|

The new `processResolvablePriceRequests` function has a maximum gas consumption of 10033176, which is slightly higher than the old version's maximum of 10000676. However, the average gas consumption of 398914 is lower than the old version's average of 408182.
requestPrice:

The new `requestPrice` function has a maximum gas consumption of 176259, which is the same as the old version's maximum of 176281. However, the average gas consumption of 160237 is lower than the old version's average of 160922.


The new `updateTrackers` function has a maximum gas consumption of 850173, which is the same as the old version's maximum of 850195. The average gas consumption of 127541 is slightly lower than the old version's average of 127484.

File: packages/core/contracts/data-verification-mechanism/implementation/VotingV2.sol (hunk 2 above)
Comment (md0x) [reply packages/core/contracts/data-verification-mechanism/implementation/VotingV2.sol]: Fixed: https://github.com/UMAprotocol/protocol/pull/4415/commits/097b280d10b086727b7ae3505badadf70219713c

**Review by chrismaree — APPROVED** [review]
this LGTM! overall slightly reduced gas sounds good and the changes make sense.

File: packages/core/contracts/data-verification-mechanism/implementation/VotingV2.sol [hunk 4]
```diff
@@ -1066,12 +1062,18 @@ contract VotingV2 is Staker, OracleInterface, OracleAncillaryInterface, OracleGo
...
+        if (priceRequest.lastVotingRound < currentRoundId) {
+            // Check if the request has already been resolved
             VoteInstance storage voteInstance = priceRequest.voteInstances[priceRequest.lastVotingRound];
             (bool isResolved, ) = _getResolvedPrice(voteInstance, priceRequest.lastVotingRound);
-            return isResolved ? RequestStatus.Resolved : RequestStatus.Active;
-        } else if (priceRequest.lastVotingRound == currentRoundId) return RequestStatus.Active;
-        else return RequestStatus.Future; // Means than priceRequest.lastVotingRound > currentRoundId
+            if (isResolved) return RequestStatus.Resolved;
+            if (_shouldDeleteRequest(_getActualRollCount(priceRequest, currentRoundId), priceRequest.isGovernance))
+                return RequestStatus.ToDelete;
```
Comment (Reinis-FRP) [inline packages/core/contracts/data-verification-mechanism/implementation/VotingV2.sol:1071]: Please also update the docstrings for _getRequestStatus to reflect possible returned state

File: packages/core/contracts/data-verification-mechanism/implementation/VotingV2.sol [hunk 5]
```diff
@@ -1116,4 +1118,20 @@ contract VotingV2 is Staker, OracleInterface, OracleAncillaryInterface, OracleGo
         RegistryInterface registry = RegistryInterface(finder.getImplementationAddress(OracleInterfaces.Registry));
         require(registry.isContractRegistered(msg.sender) || msg.sender == migratedAddress, "Caller not registered");
     }
+
+    // Checks if a request should be deleted. A request should be deleted if it has been rolled more than the maxRolls
```
Comment (Reinis-FRP) [inline packages/core/contracts/data-verification-mechanism/implementation/VotingV2.sol:1122]:
```suggestion
    // Checks if a request should be deleted. A non-gevernance request should be deleted if it has been rolled more than
    // the maxRolls
```

File: packages/core/test/hardhat/data-verification-mechanism/VotingV2.js [hunk 6]
```diff
@@ -4267,6 +4262,67 @@ describe("VotingV2", function () {
...
+
+    await moveToNextRound(voting, accounts[0]);
+
+    assert.equal((await voting.methods.getPendingRequests().call()).length, 0);
+    assert.equal((await voting.methods.getPriceRequestStatuses([{ identifier, time }]).call())[0].status, "4");
+
+    assert(
+      await didContractThrow(
+        voting.methods.getPrice(identifier, time).send({ from: account1 }),
+        "Price will be deleted"
```
Comment (Reinis-FRP) [inline packages/core/test/hardhat/data-verification-mechanism/VotingV2.js:4294]: This argument is ignored. To test specific revert messages now can use `didContractRevertWith` from common

File: packages/core/test/hardhat/data-verification-mechanism/VotingV2.js [hunk 7]
```diff
@@ -4267,6 +4262,67 @@ describe("VotingV2", function () {
...
+
+    await moveToNextRound(voting, accounts[0]);
+
+    assert.equal((await voting.methods.getPendingRequests().call()).length, 1);
+    assert.equal((await voting.methods.getPendingRequests().call())[0].rollCount, 0);
+
+    assert(
+      await didContractThrow(
+        voting.methods.getPrice(identifier, time).send({ from: account1 }),
+        "Current voting round not ended"
```
Comment (Reinis-FRP) [inline packages/core/test/hardhat/data-verification-mechanism/VotingV2.js:4322]: This argument is ignored. To test specific revert messages now can use `didContractRevertWith` from common

**Review by Reinis-FRP — APPROVED** [review]

**Pull request merged**